contracts/chain-adapters/Arbitrum_Adapter.sol

```diff
@@ -0,0 +1,124 @@
+        address _token,
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes calldata _data
+    ) external payable returns (bytes memory);
+}
+
+contract Arbitrum_Adapter is Base_Adapter {
+    // Gas limit for immediate L2 execution attempt (can be estimated via NodeInterface.estimateRetryableTicket).
+    // NodeInterface precompile interface exists at L2 address 0x00000000000000000000000000000000000000C8
+    uint32 public l2GasLimit = 5_000_000;
+
+    // Amount of ETH allocated to pay for the base submission fee. The base submission fee is a parameter unique to
+    // retryable transactions; the user is charged the base submission fee to cover the storage costs of keeping their
+    // ticket’s calldata in the retry buffer. (current base submission fee is queryable via
+    // ArbRetryableTx.getSubmissionPrice). ArbRetryableTicket precompile interface exists at L2 address
+    // 0x000000000000000000000000000000000000006E.
+    uint256 public l2MaxSubmissionCost = 0.1e18;
+
+    // L2 Gas price bid for immediate L2 execution attempt (queryable via standard eth*gasPrice RPC)
+    uint256 public l2GasPrice = 10e9; // 10 gWei
+
+    // This address on L2 receives extra ETH that is left over after relaying a message via the inbox.
+    address public l2RefundL2Address;
+
+    ArbitrumL1InboxLike public l1Inbox;
+
+    ArbitrumL1ERC20GatewayLike public l1ERC20Gateway;
+
+    event L2GasLimitSet(uint32 newL2GasLimit);
+
+    event L2MaxSubmissionCostSet(uint256 newL2MaxSubmissionCost);
+
+    event L2GasPriceSet(uint256 newL2GasPrice);
+
+    event l2RefundL2AddressSet(address newL2RefundL2Address);
+
+    constructor(
+        address _hubPool,
+        ArbitrumL1InboxLike _l1ArbitrumInbox,
+        ArbitrumL1ERC20GatewayLike _l1ERC20Gateway
+    ) Base_Adapter(_hubPool) {
+        l1Inbox = _l1ArbitrumInbox;
+        l1ERC20Gateway = _l1ERC20Gateway;
+
+        l2RefundL2Address = owner();
+    }
+
+    function setL2GasLimit(uint32 _l2GasLimit) public onlyOwner {
+        l2GasLimit = _l2GasLimit;
+        emit L2GasLimitSet(l2GasLimit);
+    }
+
+    function setL2MaxSubmissionCost(uint256 _l2MaxSubmissionCost) public onlyOwner {
+        l2MaxSubmissionCost = _l2MaxSubmissionCost;
+        emit L2MaxSubmissionCostSet(l2MaxSubmissionCost);
+    }
+
+    function setL2GasPrice(uint256 _l2GasPrice) public onlyOwner {
+        l2GasPrice = _l2GasPrice;
+        emit L2GasPriceSet(l2GasPrice);
+    }
+
+    function setL2RefundL2Address(address _l2RefundL2Address) public onlyOwner {
+        l2RefundL2Address = _l2RefundL2Address;
+        emit l2RefundL2AddressSet(l2RefundL2Address);
+    }
+
+    function relayMessage(address target, bytes memory message) external payable override onlyHubPool {
+        uint256 requiredL1CallValue = l2MaxSubmissionCost + l2GasPrice * l2GasLimit;
+        require(address(this).balance >= requiredL1CallValue, "Insufficient ETH balance");
```

Comment:
WDYT about making this required L1 call value publicly queryable, for a potential EOA's convenience? We do this with the cross-chain oracle [here](https://github.com/UMAprotocol/protocol/blob/7e0c5e049ae36a56af395d2f55c13348b5b106ac/packages/core/contracts/cross-chain-oracle/chain-adapters/Arbitrum_ParentMessenger.sol#L167) but it may not be useful for this case if the HubPool deterministically decides when to send over ETH

Reply:
ye, that's a good idea. it will also let us automate when to top up the contract. I'll refactor now.

Reply:
I dont, however, think this needs to be `nonReentrantView` like your method. that's way overkill IMHO